src/rust/engine/fs/store/src/remote/base_opendal.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+#![allow(dead_code)]
+
+use std::collections::HashSet;
+use std::time::Instant;
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::future;
+use hashing::{async_verified_copy, Digest, Fingerprint, EMPTY_DIGEST};
+use opendal::layers::{ConcurrentLimitLayer, RetryLayer, TimeoutLayer};
+use opendal::{Builder, Operator};
+use tokio::fs::File;
+use workunit_store::ObservationMetric;
+
+use super::{ByteStoreProvider, LoadDestination, RemoteOptions};
+
+#[derive(Debug, Clone, Copy)]
+pub enum LoadMode {
+  Validate,
+  NoValidate,
+}
+
+pub struct Provider {
+  /// This is public for easier testing of the action cache provider
+  // TODO: move all the providers into a single crate so that the pub isn't necessary
```

Comment:
So, on the contrary: I think that it would be good to separate the `grpc` based provider into a separate crate from the `opendal` provider, so that they can be compiled independently / in-parallel.

The `process_execution` crate had previously gotten unwieldy due to having `grpc`, `docker`, and `nailgun` clients all living in it, and I think that the `store` crate is likely to end up in the same position with another client implementation added to it (particular one as potentially-hefty as `opendal`).

Reply:
Yeah, potentially a crate for with the traits and then separate providers is better.

This comment is a bit ambiguous, but it's more referencing that it'd be nice to have the store and action cache providers for each service in the same crate, not necessarily that the gRPC/REAPI and OpenDAL ones should be grouped too.

Reply:
Getting into a bit of a distracting discussion, but separate crates for each provider might end up with 4 new crates, which is probably fine/manageable?

```mermaid
graph BT
   store("fs/store (existing)")
   remote("process_execution/remote (existing)")
   traits("remote_providers/traits (new)")
   grpc("remote_providers/reapi (new)")
   opendal("remote_providers/opendal (new)")
   selector("remote_providers (new)")
   grpc --> traits
   opendal --> traits
   selector --> grpc
   selector --> opendal
   store --> selector
   remote --> selector
```

Where the "top-level" `remote_providers` centralises things like the functions for choosing a provider based on options, and, potentially more of the validation code that's currently in `global_options.py`.

Reply:
That layout looks good, yea.